network/simulation/simulation.go

```diff
@@ -71,19 +80,66 @@ type ServiceFunc func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Se
 // every ServiceFunc must return a node.Service of the unique type.
 // This restriction is required by node.Node.Start() function
 // which is used to start node.Service returned by ServiceFunc.
-func New(services map[string]ServiceFunc) (s *Simulation) {
+func NewInProc(services map[string]ServiceFunc) (s *Simulation) {
+	s = &Simulation{
+		buckets:           make(map[enode.ID]*sync.Map),
+		done:              make(chan struct{}),
+		neighbourhoodSize: network.NewKadParams().NeighbourhoodSize,
+		typ:               SimulationTypeInproc,
+	}
+
+	s.addServices(services)
+	adapterServices := s.toAdapterServices(services)
+
+	s.Net = simulations.NewNetwork(
+		adapters.NewTCPAdapter(adapterServices),
+		&simulations.NetworkConfig{ID: "0"},
+	)
+
+	return s
+}
+
+// NewExec does the same as New but lets the caller specify the adapter to use
+func NewExec(services map[string]ServiceFunc) (s *Simulation, err error) {
 	s = &Simulation{
 		buckets:           make(map[enode.ID]*sync.Map),
 		done:              make(chan struct{}),
 		neighbourhoodSize: network.NewKadParams().NeighbourhoodSize,
+		typ:               SimulationTypeExec,
 	}
 
+	s.addServices(services)
+	adapterServices := s.toAdapterServices(services)
+
+	// exec adapters register services up front, not at node creation time
+	adapters.RegisterServices(adapterServices)
```

Comment:
@nolash  We have a problem with this `RegisterServices()` function. It should be called within an init() function as described in the documentation: https://github.com/ethereum/go-ethereum/tree/master/p2p/simulations#services 

So basically, to make the current exec/docker adapter work. we would have to register all services upfront.
In our case it gets a bit harder, because we attached this to `(s *Simulation)`... so unless we define all our simulations inside an `init()`, this won't work. 

I've been thinking now about this for a while and I don't find a decent solution.... any idea?

Reply:
I had completely forgotten about this requirement, I remember reading about it a year or so ago...

I also don't have any solutions :(

Reply:
> In our case it gets a bit harder

Why exactly do we _need_ to put it in an init?